pkg/menu/menu.go

```diff
@@ -0,0 +1,72 @@
+package menu
+
+import (
+	"io"
+
+	ui "github.com/gizak/termui/v3"
+	"github.com/gizak/termui/v3/widgets"
+)
+
+const menuWidth = 50
+
+// newParagraph will return a paragraph object with given initial text.
+func newParagraph(initText string, border bool, location int, wid int, ht int) *widgets.Paragraph {
+	p := widgets.NewParagraph()
+	p.Text = initText
+	p.Border = border
+	p.SetRect(0, location, wid, location+ht)
+	p.TextStyle.Fg = ui.ColorWhite
+	ui.Render(p)
+	return p
+}
+
+func processInput(input, warning *widgets.Paragraph, isValid func(string) (bool, string), uiEvents <-chan ui.Event) (string, error) {
+	// keep tracking all input from user
+	for {
+		e := <-uiEvents
+		if e.Type != ui.KeyboardEvent {
+			continue
+		}
+		switch e.ID {
+		// if user input control-d, raise error io.EOF meaning no more input.
+		case "<C-d>":
+			return "", io.EOF
+		case "<Enter>":
+			if isValid == nil {
+				return input.Text, nil
+			}
+			valid, warningWords := isValid(input.Text)
+			if valid {
+				return input.Text, nil
+			}
+			input.Text = ""
+			warning.Text = warningWords
+			ui.Render(input)
+			ui.Render(warning)
+		// If user hit backspace, remove the last input character
+		case "<Backspace>":
+			input.Text = input.Text[:len(input.Text)-1]
```

Comment:
If the first thing they hit is a backspace, what happens? Does input.Text[:-1] work?

see https://play.golang.org/p/Tc-3E5sEf56